src/provider/default.ts

```diff
@@ -0,0 +1,311 @@
+import axios from 'axios';
+
+import {
+  AddTransactionResponse,
+  CallContractResponse,
+  CallContractTransaction,
+  CompiledContract,
+  GetBlockResponse,
+  GetCodeResponse,
+  GetContractAddressesResponse,
+  GetTransactionResponse,
+  GetTransactionStatusResponse,
+  Transaction,
+} from '../types';
+import { parse, stringify } from '../utils/json';
+import { BigNumberish, toBN, toHex } from '../utils/number';
+import { compressProgram, formatSignature, randomAddress, wait } from '../utils/starknet';
+import { ProviderInterface } from './interface';
+
+type NetworkName = 'alpha';
+
+interface ProviderOptions {
+  network?: NetworkName;
+}
+
+export class Provider implements ProviderInterface {
+  public baseUrl: string;
+
+  public feederGatewayUrl: string;
+
+  public gatewayUrl: string;
+
+  constructor(optionsOrProvider?: ProviderOptions | Provider) {
+    if (optionsOrProvider instanceof Provider) {
+      this.baseUrl = optionsOrProvider.baseUrl;
+      this.feederGatewayUrl = optionsOrProvider.feederGatewayUrl;
+      this.gatewayUrl = optionsOrProvider.gatewayUrl;
+    } else {
+      const { network = 'alpha' } = optionsOrProvider || {};
+      const baseUrl = Provider.getNetworkFromName(network);
+      this.baseUrl = baseUrl;
+      this.feederGatewayUrl = `${baseUrl}/feeder_gateway`;
+      this.gatewayUrl = `${baseUrl}/gateway`;
+    }
+  }
+
+  protected static getNetworkFromName(name: NetworkName) {
+    switch (name) {
+      case 'alpha':
+      default:
+        return 'https://alpha3.starknet.io';
+    }
+  }
+
+  /**
+   * Gets the smart contract address on the goerli testnet.
+   *
+   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L13-L15)
+   * @returns starknet smart contract addresses
+   */
+  public getContractAddresses(): Promise<GetContractAddressesResponse> {
+    return new Promise((resolve, reject) => {
+      axios
+        .get<GetContractAddressesResponse>(`${this.feederGatewayUrl}/get_contract_addresses`)
+        .then((resp) => {
+          resolve(resp.data);
+        })
+        .catch(reject);
+    });
+  }
+
+  /**
+   * Calls a function on the StarkNet contract.
+   *
+   * [Reference](https://github.com/starkware-libs/cairo-lang/blob/f464ec4797361b6be8989e36e02ec690e74ef285/src/starkware/starknet/services/api/feeder_gateway/feeder_gateway_client.py#L17-L25)
+   *
+   * @param invokeTx - transaction to be invoked
+   * @param blockId
+   * @returns the result of the function on the smart contract.
+   */
+  public callContract(
+    invokeTx: CallContractTransaction,
```

Comment:
I think Tx is a well known abbreviation in the domain. It's also used by starknet directly. Therefore I'd like to keep this. JSDoc does the rest

Reply:
Yes better be consistent with the rest of the repo. But keep in mind ethers.js and most other libraries use the full `transaction` wording. And since Signer/Provider are concepts borrowed from ethers it depends on how much we'll want to integrate with ethers later